runtime/parachains/src/hrmp/benchmarking.rs

```diff
@@ -0,0 +1,294 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::{
+	configuration::Pallet as Configuration,
+	hrmp::{Pallet as Hrmp, *},
+	paras::Pallet as Paras,
+	shared::Pallet as Shared,
+};
+use frame_support::{assert_ok, traits::Currency};
+
+type BalanceOf<T> =
+	<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+fn register_parachain_with_balance<T: Config>(id: ParaId, balance: BalanceOf<T>) {
+	assert_ok!(Paras::<T>::initialize_para_now(
+		id,
+		crate::paras::ParaGenesisArgs {
+			parachain: true,
+			genesis_head: vec![1].into(),
+			validation_code: vec![1].into(),
+		},
+	));
+	T::Currency::make_free_balance_be(&id.into_account(), balance);
+}
+
+fn assert_last_event<T: Config>(generic_event: <T as Config>::Event) {
+	let events = frame_system::Pallet::<T>::events();
+	let system_event: <T as frame_system::Config>::Event = generic_event.into();
+	// compare to the last event record
+	let frame_system::EventRecord { event, .. } = &events[events.len() - 1];
+	assert_eq!(event, &system_event);
+}
+
+/// Enumerates the phase in the setup process of a channel between two parachains.
+enum ParachainSetupStep {
+	/// A channel open has been requested
+	Requested,
+	/// A channel open has been requested and accepted.
+	Accepted,
+	/// A channel open has been requested and accepted, and a session has passed and is now open.
+	Established,
+	/// A channel open has been requested and accepted, and a session has passed and is now
+	/// open, and now it has been requested to close down.
+	CloseRequested,
+}
+
+fn establish_para_connection<T: Config>(
+	from: u32,
+	to: u32,
+	until: ParachainSetupStep,
+) -> [(ParaId, crate::Origin); 2]
```

Comment:
Nit: You probably know what I am going to say about this.

Reply:
In fact I don't, enlighten me please.

Reply:
Use a tuple instead :)

Reply:
What is the rationale? an array of two is perfectly fine in my opinion and practically synonym to a tuple of two in my head.

Reply:
Sorry, I assumed too much.

While I think an array of two is definitely uncommon in such cases, and by default I would expect a tuple in cases like this. However, I agree it's more or less synonymous (It seems a bit novel and I even like it a bit compared to a tuple, because you use different brackets for the outer and inner items)

I, however, meant to propose to use a struct instead of the array. The inner tuples are OK since they are at least typed. That has an advantage that you fix names in one place — at the declaration side compared to the each place (which promotes consistency). Names are a bit better in this setting compared to indices: 

What's more clear to refer the recipient, a field named `recipient` or an element at index `1`?